Batching Paginated Sorts

CodeBrew28 opened the pull request:
Batches the paginated sorts by 
(1) Run the paginated sorts batched
(2) Run the paginated sorts not batched and executed serially, which is useful if we aren't making expensive operations
(3) Run the  paginated sorts not batched but with an expensive flag to spawn a goroutine per node. It detects if it is expensive if it has a ctx passed in. 

 
```
BenchmarkSortsBatched10Items                       50000             39933 ns/op           15834 B/op        246 allocs/op
BenchmarkSortsNotBatched10Items                    50000             29063 ns/op           12065 B/op        213 allocs/op
BenchmarkSortsNotBatchedExpensive10Items           50000             34620 ns/op           12385 B/op        213 allocs/op
BenchmarkSortsBatched100Items                      10000            116631 ns/op           54667 B/op        852 allocs/op
BenchmarkSortsNotBatched100Items                   10000            114231 ns/op           47156 B/op        939 allocs/op
BenchmarkSortsNotBatchedExpensive100Items          10000            161803 ns/op           50356 B/op        939 allocs/op
BenchmarkSortsBatched1000Items                      2000           1064689 ns/op          505393 B/op       7156 allocs/op
BenchmarkSortsNotBatched1000Items                   2000           1119087 ns/op          379250 B/op       8145 allocs/op
BenchmarkSortsNotBatchedExpensive1000Items          1000           1555780 ns/op          411252 B/op       8145 allocs/op
```

For reference when looking at the stats, the BenchmarkSortsNotBatchedExpensive numbers show the how long it use to take. Ideally everything will move to BenchmarkSortsBatched or BenchmarkSortsNotBatched and BenchmarkSortsNotBatchedExpensive will be deprecated

willhug commented on graphql/schemabuilder/pagination.go:
```diff
@@ -559,23 +559,53 @@ func (c *connectionContext) applySort(ctx context.Context, nodes []interface{},
-
-	for unscopedI, unscopedNode := range nodes {
-		i, node := unscopedI, unscopedNode
+	if sortField.BatchResolver != nil {
```
do we want the feature flag here as well?

wedaly commented on graphql/schemabuilder/pagination.go:
```diff
@@ -676,6 +706,14 @@ func getFuncReturnType(fn interface{}) reflect.Type {
 	return typ.Out(0)
 }
 
+func getFuncReturnTypeBatched(fn interface{}) reflect.Type {
```
is this called?  I can't find any other references to this function in the commit

wedaly commented on graphql/schemabuilder/pagination.go:
```diff
@@ -767,18 +807,47 @@ func (c *connectionContext) consumeSorts(sb *schemaBuilder, m *method, typ refle
+		var field *graphql.Field = nil
+		var err error = nil
+		if fn.BatchSortFunc != nil && fn.SortFunc != nil {
+			field, err = sb.buildBatchFunction(typ,
```
should we be using `buildBatchFunctionWithFallback` somewhere?

CodeBrew28 replied:
good catch!

wedaly commented on graphql/schemabuilder/types.go:
```diff
@@ -101,17 +101,71 @@ func BatchFilterFieldWithFallback(name string, batchFilter interface{}, filter i
+			m.SortFuncs = map[string]Sort{}
+		}
+		if filter, ok := m.SortFuncs[name]; ok {
+			panic("Sorts Filters have the same name: " + filter.Name)
```
this is a sort, not a filter?

wedaly commented on graphql/schemabuilder/types.go:
```diff
@@ -213,7 +267,7 @@ type method struct {
 
 	// Sort methods
-	SortFuncs map[string]interface{}
+	SortFuncs map[string]Sort
```
Maybe consider renaming this: it's a little misleading to call this `SortFuncs` when it actually points to a `Sort` structure rather than the function.

willhug commented on graphql/schemabuilder/pagination.go:
```diff
@@ -709,6 +747,8 @@ func getSort(typ reflect.Type) func([]sortReference, SortOrder) {
 		return sorts[reflect.Float64]
 	case reflect.String:
 		return sorts[reflect.String]
+	case reflect.Map:
```
why is map supported for sorts?

willhug commented on graphql/schemabuilder/pagination.go:
```diff
@@ -767,18 +807,47 @@ func (c *connectionContext) consumeSorts(sb *schemaBuilder, m *method, typ refle
 
 		// Build a GraphQL field for the function.
-		field, err := sb.buildFunction(typ, &method{Fn: fn})
+		var field *graphql.Field = nil
```
we can just define
```
var field *graphql.Field
var err error
```

willhug commented on graphql/schemabuilder/pagination.go:
```diff
@@ -555,27 +569,43 @@ func (c *connectionContext) applySort(ctx context.Context, nodes []interface{},
-	for unscopedI, unscopedNode := range nodes {
-		i, node := unscopedI, unscopedNode
+	if sortField.Batch && sortField.UseBatchFunc(ctx) {
 		g.Go(func() error {
```
don't need a goroutine for the batched version (there's only one field).

willhug commented on graphql/schemabuilder/pagination.go:
```diff
@@ -766,19 +796,57 @@ func (c *connectionContext) consumeSorts(sb *schemaBuilder, m *method, typ refle
-				sortableTyp,
-			)
+	for name, fn := range m.SortFuncsStruct {
+		if fn.SortFunc != nil {
```
can we put all of this into a `getFieldForSortStruct` function?  This for loop is pretty complicated. (Possible for follow-up)

willhug commented on graphql/schemabuilder/types.go:
```diff
@@ -97,21 +97,75 @@ func BatchFilterFieldWithFallback(name string, batchFilter interface{}, filter i
+}
+
+func BatchSortFieldWithFallback(name string, batchSort interface{}, sort interface{}, flag func(context.Context) bool) FieldFuncOption {
+	var sortStruct = Sort{
```
wait, should these just return `&method{` structs?  Is there any reason for us to hold onto a specific `Sort{` struct? (maybe follow-up)

willhug commented on graphql/connection_test.go:
```diff
@@ -966,3 +976,151 @@ func TestPaginatedFilters(t *testing.T) {
+	)
+
+	builtSchema := schema.MustBuild()
+	querries := []*graphql.Query{
```
arrrrrrrrrrrrr

willhug commented:
Oh, can we add a test for the Flagging logic, (e.g. batch fails, regular succeeds, flag on/off)

CodeBrew28 commented:
@willhug they're already tests I added for this in connection_test.go, I added a one more for fallback functions, but lmk if there are any edge cases you want me to test

willhug commented on graphql/schemabuilder/pagination.go:
```diff
@@ -776,12 +812,45 @@ func (c *connectionContext) consumeSorts(sb *schemaBuilder, m *method, typ refle
 		c.SortFunctions[name] = getSort(sortableTyp)
+	}
+
+	if sortMethod.Batch == true {
```
if/else?
